### PR TITLE
Use .env pronoun (fixes #382)

### LINKED ISSUE #382: `finalize_estimator` fails in metric summarizers when any column is named `name`
## The problem

If any column of a data frame passed to `data` is named `name`, `finalize_estimator` will fail. This seems to be happening because `name` is being evaluated in the context of `data` inside of `summarise`:

https://github.com/tidymodels/yardstick/blob/3099e9973e0983ff134406d18920f82a8913dc32/R/template.R#L119-L120

Changing both of those lines to use `.env[["name"]]` seems to fix this behavior:

``` r
Orange2 <- Orange
Orange2$name <- Orange$circumference
library(rlang)
assignInNamespace(
  "numeric_metric_summarizer", 
  function(name,
           fn,
           data,
           truth,
           estimate,
           ...,
           na_rm = TRUE,
           case_weights = NULL,
           fn_options = list(),
           error_call = caller_env()) {
    rlang::check_dots_empty()
    
    truth <- enquo(truth)
    estimate <- enquo(estimate)
    case_weights <- enquo(case_weights)
    
    truth <- yardstick:::yardstick_eval_select(
      expr = truth,
      data = data,
      arg = "truth",
      error_call = error_call
    )
    estimate <- yardstick:::yardstick_eval_select(
      expr = estimate,
      data = data,
      arg = "estimate",
      error_call = error_call
    )
    
    if (!quo_is_null(case_weights)) {
      case_weights <- yardstick:::yardstick_eval_select(
        expr = case_weights,
        data = data,
        arg = "case_weights",
        error_call = error_call
      )
      
      case_weights <- expr(.data[[!!case_weights]])
    }
    
    out <- dplyr::summarise(
      data,
      .metric = .env[["name"]], # change 1
      .estimator = yardstick::finalize_estimator(
        .data[[truth]], 
        metric_class = .env[["name"]]), # change 2
      .estimate = fn(
        truth = .data[[truth]],
        estimate = .data[[estimate]],
        case_weights = !!case_weights,
        na_rm = na_rm,
        !!!fn_options
      )
    )
    
    dplyr::as_tibble(out)
  },
  "yardstick"
)
yardstick::rmse(Orange2, name, circumference)
#> # A tibble: 1 × 3
#>   .metric .estimator .estimate
#>   <chr>   <chr>          <dbl>
#> 1 rmse    standard           0
```

<sup>Created on 2023-03-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.2 Patched (2022-11-10 r83330)
#>  os       Ubuntu 22.04.2 LTS
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/New_York
#>  date     2023-03-02
#>  pandoc   2.19.2 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date (UTC) lib source
#>  cli           3.6.0      2023-01-09 [1] CRAN (R 4.2.2)
#>  digest        0.6.31     2022-12-11 [1] CRAN (R 4.2.2)
#>  dplyr         1.1.0      2023-01-29 [1] CRAN (R 4.2.2)
#>  evaluate      0.20       2023-01-17 [1] CRAN (R 4.2.2)
#>  fansi         1.0.4      2023-01-22 [1] CRAN (R 4.2.2)
#>  fastmap       1.1.1      2023-02-24 [1] CRAN (R 4.2.2)
#>  fs            1.6.0      2023-01-23 [1] CRAN (R 4.2.2)
#>  generics      0.1.3      2022-07-05 [1] CRAN (R 4.2.2)
#>  glue          1.6.2      2022-02-24 [1] CRAN (R 4.2.2)
#>  htmltools     0.5.4      2022-12-07 [1] CRAN (R 4.2.2)
#>  knitr         1.42       2023-01-25 [1] CRAN (R 4.2.2)
#>  lifecycle     1.0.3      2022-10-07 [1] CRAN (R 4.2.2)
#>  magrittr      2.0.3      2022-03-30 [1] CRAN (R 4.2.2)
#>  pillar        1.8.1      2022-08-19 [1] CRAN (R 4.2.2)
#>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.2.2)
#>  purrr         1.0.1      2023-01-10 [1] CRAN (R 4.2.2)
#>  R.cache       0.16.0     2022-07-21 [1] CRAN (R 4.2.2)
#>  R.methodsS3   1.8.2      2022-06-13 [1] CRAN (R 4.2.2)
#>  R.oo          1.25.0     2022-06-12 [1] CRAN (R 4.2.2)
#>  R.utils       2.12.2     2022-11-11 [1] CRAN (R 4.2.2)
#>  R6            2.5.1      2021-08-19 [1] CRAN (R 4.2.2)
#>  reprex        2.0.2      2022-08-17 [1] CRAN (R 4.2.2)
#>  rlang       * 1.0.6      2022-09-24 [1] CRAN (R 4.2.2)
#>  rmarkdown     2.20       2023-01-19 [1] CRAN (R 4.2.2)
#>  rstudioapi    0.14       2022-08-22 [1] CRAN (R 4.2.2)
#>  sessioninfo   1.2.2      2021-12-06 [1] CRAN (R 4.2.2)
#>  styler        1.8.1      2022-11-07 [1] CRAN (R 4.2.2)
#>  tibble        3.1.8      2022-07-22 [1] CRAN (R 4.2.2)
#>  tidyselect    1.2.0      2022-10-10 [1] CRAN (R 4.2.2)
#>  utf8          1.2.3      2023-01-31 [1] CRAN (R 4.2.2)
#>  vctrs         0.5.2      2023-01-23 [1] CRAN (R 4.2.2)
#>  withr         2.5.0      2022-03-03 [1] CRAN (R 4.2.2)
#>  xfun          0.36       2022-12-21 [1] CRAN (R 4.2.2)
#>  yaml          2.3.7      2023-01-23 [1] CRAN (R 4.2.2)
#>  yardstick     1.1.0.9000 2023-03-02 [1] Github (tidymodels/yardstick@3099e99)
#> 
#>  [1] /home/mikemahoney218/R/x86_64-pc-linux-gnu-library/4.2
#>  [2] /usr/local/lib/R/site-library
#>  [3] /usr/lib/R/site-library
#>  [4] /usr/lib/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

I only tested with numeric metrics but I believe this is the same for the other summarizers as well.

## Reproducible example
``` r
Orange2 <- Orange
Orange2$name <- Orange$circumference
yardstick::rmse(Orange2, circumference, name)
#> Error in `dplyr::summarise()`:
#> ℹ In argument: `.estimator =
#>   finalize_estimator(.data[["circumference"]], metric_class = name)`.
#> Caused by error in `attributes(.Data) <- c(attributes(.Data), attrib)`:
#> ! attempt to set invalid 'class' attribute

#> Backtrace:
#>      ▆
#>   1. ├─yardstick::rmse(Orange2, circumference, name)
#>   2. ├─yardstick:::rmse.data.frame(Orange2, circumference, name)
#>   3. │ └─yardstick::numeric_metric_summarizer(...)
#>   4. │   ├─dplyr::summarise(...)
#>   5. │   └─dplyr:::summarise.data.frame(...)
#>   6. │     └─dplyr:::summarise_cols(.data, dplyr_quosures(...), by, "summarise")
#>   7. │       ├─base::withCallingHandlers(...)
#>   8. │       └─dplyr:::map(quosures, summarise_eval_one, mask = mask)
#>   9. │         └─base::lapply(.x, .f, ...)
#>  10. │           └─dplyr (local) FUN(X[[i]], ...)
#>  11. │             └─mask$eval_all_summarise(quo)
#>  12. │               └─dplyr (local) eval()
#>  13. ├─yardstick::finalize_estimator(.data[["circumference"]], metric_class = name)
#>  14. │ └─yardstick:::make_dummy(metric_class)
#>  15. │   └─base::structure(list(), class = metric_class)
#>  16. └─base::.handleSimpleError(...)
#>  17.   └─dplyr (local) h(simpleError(msg, call))
#>  18.     └─dplyr (local) handler(cnd)
#>  19.       └─rlang::abort(message, class = error_class, parent = parent, call = error_call)
yardstick::rmse(Orange2, name, circumference)
#> Error in `dplyr::summarise()`:
#> ℹ In argument: `.estimator = finalize_estimator(.data[["name"]],
#>   metric_class = name)`.
#> Caused by error in `attributes(.Data) <- c(attributes(.Data), attrib)`:
#> ! attempt to set invalid 'class' attribute

#> Backtrace:
#>      ▆
#>   1. ├─yardstick::rmse(Orange2, name, circumference)
#>   2. ├─yardstick:::rmse.data.frame(Orange2, name, circumference)
#>   3. │ └─yardstick::numeric_metric_summarizer(...)
#>   4. │   ├─dplyr::summarise(...)
#>   5. │   └─dplyr:::summarise.data.frame(...)
#>   6. │     └─dplyr:::summarise_cols(.data, dplyr_quosures(...), by, "summarise")
#>   7. │       ├─base::withCallingHandlers(...)
#>   8. │       └─dplyr:::map(quosures, summarise_eval_one, mask = mask)
#>   9. │         └─base::lapply(.x, .f, ...)
#>  10. │           └─dplyr (local) FUN(X[[i]], ...)
#>  11. │             └─mask$eval_all_summarise(quo)
#>  12. │               └─dplyr (local) eval()
#>  13. ├─yardstick::finalize_estimator(.data[["name"]], metric_class = name)
#>  14. │ └─yardstick:::make_dummy(metric_class)
#>  15. │   └─base::structure(list(), class = metric_class)
#>  16. └─base::.handleSimpleError(...)
#>  17.   └─dplyr (local) h(simpleError(msg, call))
#>  18.     └─dplyr (local) handler(cnd)
#>  19.       └─rlang::abort(message, class = error_class, parent = parent, call = error_call)
# "name" doesn't need to be passed to the function:
yardstick::rmse(Orange2, age, circumference)
#> Error in `dplyr::summarise()`:
#> ℹ In argument: `.estimator = finalize_estimator(.data[["age"]],
#>   metric_class = name)`.
#> Caused by error in `attributes(.Data) <- c(attributes(.Data), attrib)`:
#> ! attempt to set invalid 'class' attribute

#> Backtrace:
#>      ▆
#>   1. ├─yardstick::rmse(Orange2, age, circumference)
#>   2. ├─yardstick:::rmse.data.frame(Orange2, age, circumference)
#>   3. │ └─yardstick::numeric_metric_summarizer(...)
#>   4. │   ├─dplyr::summarise(...)
#>   5. │   └─dplyr:::summarise.data.frame(...)
#>   6. │     └─dplyr:::summarise_cols(.data, dplyr_quosures(...), by, "summarise")
#>   7. │       ├─base::withCallingHandlers(...)
#>   8. │       └─dplyr:::map(quosures, summarise_eval_one, mask = mask)
#>   9. │         └─base::lapply(.x, .f, ...)
#>  10. │           └─dplyr (local) FUN(X[[i]], ...)
#>  11. │             └─mask$eval_all_summarise(quo)
#>  12. │               └─dplyr (local) eval()
#>  13. ├─yardstick::finalize_estimator(.data[["age"]], metric_class = name)
#>  14. │ └─yardstick:::make_dummy(metric_class)
#>  15. │   └─base::structure(list(), class = metric_class)
#>  16. └─base::.handleSimpleError(...)
#>  17.   └─dplyr (local) h(simpleError(msg, call))
#>  18.     └─dplyr (local) handler(cnd)
#>  19.       └─rlang::abort(message, class = error_class, parent = parent, call = error_call)
```

<sup>Created on 2023-03-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.2 Patched (2022-11-10 r83330)
#>  os       Ubuntu 22.04.2 LTS
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/New_York
#>  date     2023-03-02
#>  pandoc   2.19.2 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date (UTC) lib source
#>  cli           3.6.0      2023-01-09 [1] CRAN (R 4.2.2)
#>  digest        0.6.31     2022-12-11 [1] CRAN (R 4.2.2)
#>  dplyr         1.1.0      2023-01-29 [1] CRAN (R 4.2.2)
#>  evaluate      0.20       2023-01-17 [1] CRAN (R 4.2.2)
#>  fansi         1.0.4      2023-01-22 [1] CRAN (R 4.2.2)
#>  fastmap       1.1.1      2023-02-24 [1] CRAN (R 4.2.2)
#>  fs            1.6.0      2023-01-23 [1] CRAN (R 4.2.2)
#>  generics      0.1.3      2022-07-05 [1] CRAN (R 4.2.2)
#>  glue          1.6.2      2022-02-24 [1] CRAN (R 4.2.2)
#>  htmltools     0.5.4      2022-12-07 [1] CRAN (R 4.2.2)
#>  knitr         1.42       2023-01-25 [1] CRAN (R 4.2.2)
#>  lifecycle     1.0.3      2022-10-07 [1] CRAN (R 4.2.2)
#>  magrittr      2.0.3      2022-03-30 [1] CRAN (R 4.2.2)
#>  pillar        1.8.1      2022-08-19 [1] CRAN (R 4.2.2)
#>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.2.2)
#>  purrr         1.0.1      2023-01-10 [1] CRAN (R 4.2.2)
#>  R.cache       0.16.0     2022-07-21 [1] CRAN (R 4.2.2)
#>  R.methodsS3   1.8.2      2022-06-13 [1] CRAN (R 4.2.2)
#>  R.oo          1.25.0     2022-06-12 [1] CRAN (R 4.2.2)
#>  R.utils       2.12.2     2022-11-11 [1] CRAN (R 4.2.2)
#>  R6            2.5.1      2021-08-19 [1] CRAN (R 4.2.2)
#>  reprex        2.0.2      2022-08-17 [1] CRAN (R 4.2.2)
#>  rlang         1.0.6      2022-09-24 [1] CRAN (R 4.2.2)
#>  rmarkdown     2.20       2023-01-19 [1] CRAN (R 4.2.2)
#>  rstudioapi    0.14       2022-08-22 [1] CRAN (R 4.2.2)
#>  sessioninfo   1.2.2      2021-12-06 [1] CRAN (R 4.2.2)
#>  styler        1.8.1      2022-11-07 [1] CRAN (R 4.2.2)
#>  tibble        3.1.8      2022-07-22 [1] CRAN (R 4.2.2)
#>  tidyselect    1.2.0      2022-10-10 [1] CRAN (R 4.2.2)
#>  utf8          1.2.3      2023-01-31 [1] CRAN (R 4.2.2)
#>  vctrs         0.5.2      2023-01-23 [1] CRAN (R 4.2.2)
#>  withr         2.5.0      2022-03-03 [1] CRAN (R 4.2.2)
#>  xfun          0.36       2022-12-21 [1] CRAN (R 4.2.2)
#>  yaml          2.3.7      2023-01-23 [1] CRAN (R 4.2.2)
#>  yardstick     1.1.0.9000 2023-03-02 [1] Github (tidymodels/yardstick@3099e99)
#> 
#>  [1] /home/mikemahoney218/R/x86_64-pc-linux-gnu-library/4.2
#>  [2] /usr/local/lib/R/site-library
#>  [3] /usr/lib/R/site-library
#>  [4] /usr/lib/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

### PR DESCRIPTION
**Pull request by mikemahoney218**
This PR fixes #382 by using the .env pronoun to select "name" and "estimator" from the local environment, rather than the data mask, when appropriate. Figured it was an easy enough fix that I might as well send in a PR :smile: 

The tests also add an "event_level" column because I couldn't guess without checking if the `!!! spliceable_argument(event_level, "event_level")` sections also needed to use the pronoun (they do not, and cannot use it).

### CODE REVIEW

**Comment by EmilHvitfeldt** [comment]
I'm tagging @DavisVaughan in because this is much more his domain (land of rlang)

**Comment by DavisVaughan** [comment]
Does it also fail if a column is named `na_rm`?

**Comment by mikemahoney218** [comment]
Not anymore :laughing: Thanks for the catch!

File: R/template.R [hunk 1]
```diff
@@ -233,16 +240,20 @@ prob_metric_summarizer <- function(name,
...
+    .metric = .env[["name"]],
+    .estimator = finalize_estimator(
+      .data[[truth]],
+      .env[["estimator"]],
+      .env[["name"]]
+    ),
     .estimate = fn(
       truth = .data[[truth]],
       estimate = {
         estimate <- dplyr::pick(tidyselect::all_of(estimate))
```
Comment (DavisVaughan) [inline R/template.R:252]:
```suggestion
        estimate <- dplyr::pick(tidyselect::all_of(.env[["estimate"]]))
```

Check me on this one, maybe add a test for it

File: R/template.R [hunk 2]
```diff
@@ -293,16 +304,20 @@ curve_metric_summarizer <- function(name,
...
```
Comment (DavisVaughan) [inline R/template.R:316]: And here, with a test

**Review by DavisVaughan — APPROVED** [review]
Looking great, so I'll approve. Just one last question below

File: R/template.R [hunk 3]
```diff
@@ -116,13 +116,16 @@ numeric_metric_summarizer <- function(name,
```
Comment (DavisVaughan) [inline R/template.R:116]: Oh and add a NEWS bullet mentioning which issue it closes?

**Comment by mikemahoney218** [comment]
Great! I'm OOO so will push tests and a news bullet next week 😄

**Comment by mikemahoney218** [comment]
I added tests for both `truth` and `estimate` as column names to avoid regressions, and added a bullet to NEWS. Thanks!

**Comment by EmilHvitfeldt** [comment]
Amazing! Thank you both!

**Pull request merged**